src/Components/Components/src/Routing/RouteTableFactory.cs

```diff
@@ -47,57 +58,87 @@ private static List<Type> GetRouteableComponents(IEnumerable<Assembly> assemblie
                     }
                 }
             }
-
-            return routeableComponents;
         }
 
         internal static RouteTable Create(List<Type> componentTypes)
         {
-            var templatesByHandler = new Dictionary<Type, string[]>();
+            var templatesByHandler = new Dictionary<Type, List<string>>();
             foreach (var componentType in componentTypes)
             {
                 // We're deliberately using inherit = false here.
                 //
                 // RouteAttribute is defined as non-inherited, because inheriting a route attribute always causes an
                 // ambiguity. You end up with two components (base class and derived class) with the same route.
                 var routeAttributes = componentType.GetCustomAttributes<RouteAttribute>(inherit: false);
+                var templates = new List<string>();
+                foreach (var attribute in routeAttributes)
+                {
+                    templates.Add(attribute.Template);
+                }
 
-                var templates = routeAttributes.Select(t => t.Template).ToArray();
                 templatesByHandler.Add(componentType, templates);
             }
             return Create(templatesByHandler);
         }
 
-        internal static RouteTable Create(Dictionary<Type, string[]> templatesByHandler)
+        internal static RouteTable Create(Dictionary<Type, List<string>> templatesByHandler)
         {
             var routes = new List<RouteEntry>();
-            foreach (var keyValuePair in templatesByHandler)
+            foreach (var (type, templates) in templatesByHandler)
```

Comment:
How likely is it that templates is empty?  If there's a reasonable chance, you could guard the whole body of this loop to avoid all the work/allocations below if it is.